query/common/src/main/java/org/codice/ddf/admin/common/fields/base/BaseFunctionField.java

```diff
@@ -0,0 +1,129 @@
+import org.codice.ddf.admin.api.report.FunctionReport;
+import org.codice.ddf.admin.api.report.Message;
+import org.codice.ddf.admin.api.report.Report;
+import org.codice.ddf.admin.common.report.FunctionReportImpl;
+
+import com.google.common.collect.ImmutableList;
+
+public abstract class BaseFunctionField<T extends DataType> extends BaseField<Map<String, Object>, FunctionReport<T>> implements FunctionField<T> {
+
+    private FunctionReportImpl<T> report;
+
+    private T returnType;
+
+    public BaseFunctionField(String functionName, String description, T returnType) {
+        super(functionName, description);
+        this.returnType = returnType;
+        report = new FunctionReportImpl<>();
+    }
+
+    public abstract T performFunction();
+
+    @Override
+    public T getReturnType() {
+        return returnType;
+    }
+
+    @Override
+    public void setValue(Map<String, Object> args) {
+        if (args == null || args.isEmpty()) {
+            return;
+        }
+
+        getArguments().stream()
+                .filter(field -> args.containsKey(field.fieldName()))
+                .forEach(field -> field.setValue(args.get(field.fieldName())));
+    }
+
+    @Override
+    public FunctionReport<T> getValue() {
+        validate();
+        if (!report.containsErrorMsgs()) {
+            report.result(performFunction());
+        }
+
+        return report;
+    }
+
+    @Override
+    public void updatePath(List<String> subPath) {
+        super.updatePath(subPath);
+        updateArgumentPaths();
+    }
+
+    @Override
+    public void fieldName(String fieldName) {
+        super.fieldName(fieldName);
+        updateArgumentPaths();
+    }
+
+    public void updateArgumentPaths() {
+        List<String> argPath = new ImmutableList.Builder<String>().addAll(path())
+                .add(ARGUMENT)
+                .build();
+        getArguments().forEach(arg -> arg.updatePath(argPath));
+    }
+
+    public void validate() {
+        getArguments().stream()
+                .map(DataType::validate)
+                .flatMap(Collection<Message>::stream)
+                .forEach(msg -> addArgumentMessage(msg));
+    }
+
+    protected boolean containsErrorMsgs() {
+        return report.containsErrorMsgs();
+    }
+
+    protected BaseFunctionField addArgumentMessages(List<Message> msgs) {
+        msgs.forEach(msg -> addArgumentMessage(msg));
+        return this;
+    }
+
+    protected BaseFunctionField addArgumentMessage(Message msg) {
+        report.addArgumentMessage(msg);
+        return this;
+    }
+
+    protected BaseFunctionField addResultMessages(List<Message> msgs) {
+        msgs.forEach(this::addResultMessage);
+        return this;
+    }
+
+    protected BaseFunctionField addMessages(Report report) {
+        this.report.addMessages(report);
+        return this;
+    }
+
+    protected BaseFunctionField addResultMessage(Message msg) {
+        Message copy = msg.copy();
```

Comment:
I removed the `addSubpath` method, good catch! I'd like to keep the `setPath` method though, I don't see anything wrong with being able to set the path to a message, especially if the message path is empty.